hivemind/client/optim/collaborative.py

```diff
@@ -0,0 +1,260 @@
+        self.performance_ema = PerformanceEMA(alpha=performance_ema_alpha)
+
+        self.training_progress_key = f"{self.prefix}_progress"
+        self.local_samples_accumulated = 0  # a number of local samples accumulated since last optimizer update
+        self.local_steps_accumulated = 0  # a number of calls to step() since last optimizer update
+
+        self.collaboration_state = self.fetch_collaboration_state()
+        self.lock_collaboration_state, self.state_updated = Lock(), Event()
+        self.collaboration_state_updater = Thread(target=self.check_collaboration_state_periodically, daemon=True,
+                                                  name=f"{self}.collaboration_state_updater")
+        self.collaboration_state_updater.start()
+
+    @property
+    def local_step(self) -> int:
+        return self.averager.local_step
+
+    @property
+    def is_synchronized(self) -> bool:
+        return self.local_step >= self.collaboration_state.optimizer_step
+
+    def is_alive(self) -> bool:
+        return self.averager.is_alive()
+
+    def load_state_from_peers(self, **kwargs):
+        """ Attempt to fetch the newest collaboration state from other peers """
+        with self.lock_collaboration_state:
+            self.averager.load_state_from_peers(**kwargs)
+            self.local_samples_accumulated = self.local_steps_accumulated = 0
+            self.update_scheduler()
+            self.opt.zero_grad()
+
+    def step(self, batch_size: Optional[int] = None, **kwargs):
+        batch_size = self.batch_size_per_step if batch_size is None else batch_size
+        if not self.is_synchronized:
+            self.load_state_from_peers()
+            return
+
+        self.local_samples_accumulated += batch_size
+        self.local_steps_accumulated += 1
+        self.performance_ema.update(num_processed=self.batch_size_per_step)
+        run_in_background(self.report_training_progress)
+        if not self.collaboration_state.ready_for_step:
+            return
+
+        logger.info("Averaging parameters and gradients with peers...")
+        self.collaboration_state = self.fetch_collaboration_state()
+        self.state_updated.set()
+
+        if not self.is_synchronized:
+            self.load_state_from_peers()
+            return
+
+        with self.performance_ema.pause(), self.lock_collaboration_state:
+            if self.collaboration_state.num_peers > 1:
+                mean_samples_per_worker = self.target_batch_size / self.collaboration_state.num_peers
+                weight = self.local_samples_accumulated / mean_samples_per_worker
+                self.averager.step(weight=weight, timeout=self.averaging_timeout, **kwargs)
+            else:
+                logger.info(
+                    f"Skipped averaging: collaboration consists of {self.collaboration_state.num_peers} peer(s).")
+                self.averager.local_step += 1
+
+            output = self.opt.step()
+            self.opt.zero_grad()
```

Comment:
Is it true that we ignore all gradient accumulation and LR scheduling in external code? If so, we should make it very clear in the docstring

Reply:
added a better description, will create an issue for a "safe" option